File: VB/DevExpress.XtraReports.CustomControls/XRGrid.vb
Imports System.ComponentModel
Imports System.Drawing.Design
Imports DevExpress.Utils.Serializing
Imports DevExpress.XtraReports.Native.Presenters

Namespace DevExpress.XtraReports.CustomControls
	<ToolboxItem(True), Designer("DevExpress.XtraReports.CustomControls.XRGridDesigner, DevExpress.XtraReports.CustomControls"), XRDesigner("DevExpress.XtraReports.CustomControls.XRGridDesigner, DevExpress.XtraReports.CustomControls")>
	Public Class XRGrid
		Inherits XRTableLikeContainerControl

		Public Sub New()
			MyBase.New()
			WidthF = 300F
			HeightF = 200F
		End Sub

		Protected Overrides Function CreateContainerBrick(ByVal owner As XRDataContainerControl, ByVal isHeader As Boolean) As DataContainerBrick
			Return New GridBrick(owner, isHeader)
		End Function

		Protected Friend Overrides Function CreateHeader() As XRFieldHeader
			Return New XRGridColumn()
		End Function

		Protected Overrides Function CreateHeaders() As XRFieldHeaderCollection
			Return New XRGridColumnCollection(Me)
		End Function

		Protected Overrides Function CreatePresenter() As Native.Presenters.ControlPresenter
			Return MyBase.CreatePresenter(Of ControlPresenter)(Function()
				Return New XRGridRuntimePresenter(Me)
			End Function, Function()
				Return New XRGridDesignTimePresenter(Me)
			End Function)
		End Function


		<XtraSerializableProperty(XtraSerializationVisibility.Collection, True, False, False, -1, XtraSerializationFlags.Cached), DesignerSerializationVisibility(DesignerSerializationVisibility.Content)>
		Public ReadOnly Property Columns() As XRGridColumnCollection
			Get
				Return TryCast(MyBase.Headers, XRGridColumnCollection)
			End Get
		End Property

		Friend Overrides ReadOnly Property FieldHeaderName() As String
			Get
				Return "Columns"
			End Get
		End Property
	End Class

	Public Class XRGridColumn
		Inherits XRResizableFieldHeader

	End Class

	Public Class XRGridColumnCollection
		Inherits XRFieldHeaderCollection

		Public Sub New(ByVal control As XRGrid)
			MyBase.New(control)
		End Sub

		Default Public Shadows ReadOnly Property Item(ByVal fieldName As String) As XRGridColumn
			Get
				Return TryCast(MyBase.Item(fieldName), XRGridColumn)
			End Get
		End Property
		Default Public Shadows ReadOnly Property Item(ByVal index As Integer) As XRGridColumn
			Get
				Return TryCast(MyBase.Item(index), XRGridColumn)
			End Get
		End Property
	End Class
End Namespace

File: CS/DevExpress.XtraReports.CustomControls/XRGrid.cs
using System.ComponentModel;
using System.Drawing.Design;
using DevExpress.Utils.Serializing;
using DevExpress.XtraReports.Native.Presenters;

namespace DevExpress.XtraReports.CustomControls {
    [ToolboxItem(true),
    Designer("DevExpress.XtraReports.CustomControls.XRGridDesigner, DevExpress.XtraReports.CustomControls"),
    XRDesigner("DevExpress.XtraReports.CustomControls.XRGridDesigner, DevExpress.XtraReports.CustomControls")]
    public class XRGrid : XRTableLikeContainerControl
    {
        public XRGrid() : base()
        {
            WidthF = 300f;
            HeightF = 200f;
        }

        protected override DataContainerBrick CreateContainerBrick(XRDataContainerControl owner, bool isHeader)
        {
            return new GridBrick(owner, isHeader);
        }

        protected internal override XRFieldHeader CreateHeader()
        {
            return new XRGridColumn();
        }

        protected override XRFieldHeaderCollection CreateHeaders()
        {
            return new XRGridColumnCollection(this);
        }        

        protected override Native.Presenters.ControlPresenter CreatePresenter()
        {            
            return base.CreatePresenter<ControlPresenter>(delegate
            {
                return new XRGridRuntimePresenter(this);
            }, delegate
            {
                return new XRGridDesignTimePresenter(this);
            });
        }
   

        [
        XtraSerializableProperty(XtraSerializationVisibility.Collection, true, false, false, -1, XtraSerializationFlags.Cached),
        DesignerSerializationVisibility(DesignerSerializationVisibility.Content)
        ]
        public XRGridColumnCollection Columns { get { return base.Headers as XRGridColumnCollection; } }

        internal override string FieldHeaderName { get { return "Columns"; } }
    }

    public class XRGridColumn : XRResizableFieldHeader { }

    public class XRGridColumnCollection : XRFieldHeaderCollection
    {
        public XRGridColumnCollection(XRGrid control) : base(control) { }

        public new XRGridColumn this[string fieldName] { get { return base[fieldName] as XRGridColumn; } }
        public new XRGridColumn this[int index] { get { return base[index] as XRGridColumn; } }
    }
}
